spec/ddtrace/trace_operation_spec.rb

```diff
@@ -0,0 +1,161 @@
+# typed: ignore
+require 'spec_helper'
+require 'ddtrace/trace_operation'
+
+require 'ddtrace/transport/serializable_trace'
+
+RSpec.describe Datadog::TraceOperation do
+  subject(:trace_op) { described_class.new(**options) }
+  let(:options) { {} }
+
+  context 'for a mock job with fan-out/fan-in behavior' do
+    subject(:trace) do
+      trace_op.measure('job', resource: 'import_job', service: 'job-worker') do |_span, trace|
+        trace.measure('load_data', resource: 'imports.csv') do
+          trace.measure('read_file', resource: 'imports.csv') do
+            sleep(0.01)
+          end
+
+          trace.measure('deserialize', resource: 'inventory') do
+            sleep(0.01)
+          end
+        end
+
+        workers = nil
+        trace.measure('start_inserts', resource: 'inventory') do
+          trace_digest = trace.to_digest
+
+          @thread_traces = []
+          workers = Array.new(5) do |index|
+            Thread.new do
+              # Delay start-up slightly
+              sleep(0.01)
+
+              thread_trace = described_class.new(
+                id: trace_digest.trace_id,
+                origin: trace_digest.trace_origin,
+                parent_span_id: trace_digest.span_id,
+                sampling_priority: trace_digest.trace_sampling_priority
+              )
+
+              @thread_traces << thread_trace
+
+              thread_trace.measure(
+                'db.query',
+                service: 'database',
+                resource: "worker #{index}"
+              ) do
+                sleep(0.01)
+              end
+            end
+          end
+        end
+
+        trace.measure('wait_inserts', resource: 'inventory') do |wait_span|
+          wait_span.set_tag('worker.count', workers.length)
+          workers && workers.each(&:join)
+        end
+
+        trace.measure('update_log', resource: 'inventory') do
+          sleep(0.01)
+        end
+      end
+
+      trace_op.flush!
+    end
+
+    before { skip 'Thread behavior unstable on JRuby.' if PlatformHelpers.jruby? }
```

Comment:
I'm curious about this skip -- what issues did you see? Usually thread issues on JRuby exist in MRI as well, they just don't trigger as often. So this could just be uncovering a real production bug or a flaky spec.

(Same note applies to similar skips in this PR)

Reply:
The test creates 5 threads to fan-out, adds `TraceOperation` references from within the threads to an `Array` that was created outside the threads, then uses `Thread#join` to fan-in then finish the trace. It then collects all the spans from all the traces and does its assertions. In JRuby, the traces from the thread is sometimes `nil`. I don't know why, but it seems to randomly fail/pass. It passes 100% of the time on MRI.

Reply:
Looking at the test code, appending to `@thread_traces` definitely seems like a possible problem, since we don't synchronize the writes. Would suggest replacing with a `Queue` and seeing if that is the source of the issues.

If it's not that... my concern is that it may be an actual problem our customers would face.

Reply:
I'm not concerned this is a real problem: `TraceOperation` doesn't actually deal with threads, it's just a struct. Any thread issues are definitely a facet of the test itself.

After more testing, it looks like it's just flaky on JRuby, because JRuby is really bad in general with threading in tests. I can get it to pass without changes if I run it enough times. Switching to `Queue` seems to improve consistency for JRuby.

Reply:
> After more testing, it looks like it's just flaky on JRuby, because JRuby is really bad in general with threading in tests.

Note that I've used quite a bit of JRuby in the past, and my experience is that almost always it's our Ruby code than JRuby being buggy.

I'm somewhat... uncomfortable about just skipping threading-related tests on JRuby, because we do have it as supported platform. If the tests are important enough to exist, should we not have them running and passing for JRuby as well?

Reply:
It's fixed.

Reply:
Awesome, thanks for checking! 👍